src/core/lib/security/security_connector/tls/spiffe_security_connector.cc

```diff
@@ -0,0 +1,465 @@
+                  grpc_closure* on_peer_checked) override {
+    const char* target_name = overridden_target_name_ != nullptr
+                                  ? overridden_target_name_
+                                  : target_name_;
+    grpc_error* error = grpc_ssl_check_peer(target_name, &peer, auth_context);
+    const grpc_tls_spiffe_credentials* creds =
+        static_cast<const grpc_tls_spiffe_credentials*>(channel_creds());
+    GPR_ASSERT(creds->options() != nullptr);
+    const grpc_tls_server_authorization_check_config* config =
+        creds->options()->server_authorization_check_config();
+    if (error == GRPC_ERROR_NONE && config != nullptr) {
+      /* Peer property will contain a complete certificate chain. */
+      const tsi_peer_property* p =
+          tsi_peer_get_property_by_name(&peer, TSI_X509_PEM_CERT_PROPERTY);
+      if (p == nullptr) {
+        error = GRPC_ERROR_CREATE_FROM_STATIC_STRING(
+            "Cannot check peer: missing pem cert property.");
+      } else {
+        char* peer_pem = static_cast<char*>(gpr_malloc(p->value.length + 1));
+        memcpy(peer_pem, p->value.data, p->value.length);
+        peer_pem[p->value.length] = '\0';
+        GPR_ASSERT(check_arg_ != nullptr);
+        check_arg_->peer_cert = check_arg_->peer_cert == nullptr
+                                    ? gpr_strdup(peer_pem)
+                                    : check_arg_->peer_cert;
+        check_arg_->target_name = check_arg_->target_name == nullptr
+                                      ? gpr_strdup(target_name)
+                                      : check_arg_->target_name;
+        on_peer_checked_ = on_peer_checked;
+        gpr_free(peer_pem);
+        int callback_status = config->Schedule(check_arg_);
+        /* Server authorization check is handled asynchronously. */
+        if (callback_status) {
+          tsi_peer_destruct(&peer);
+          return;
+        }
+        /* Server authorization check is handled synchronously. */
+        error = ProcessServerAuthorizationCheckResult(check_arg_);
+      }
+    }
+    GRPC_CLOSURE_SCHED(on_peer_checked, error);
+    tsi_peer_destruct(&peer);
+  }
+
+  int cmp(const grpc_security_connector* other_sc) const override {
+    auto* other =
+        reinterpret_cast<const grpc_tls_spiffe_channel_security_connector*>(
+            other_sc);
+    int c = channel_security_connector_cmp(other);
+    if (c != 0) {
+      return c;
+    }
+    c = strcmp(target_name_, other->target_name_);
+    if (c != 0) {
+      return c;
+    }
+    return (overridden_target_name_ == nullptr ||
+            other->overridden_target_name_ == nullptr)
+               ? GPR_ICMP(overridden_target_name_,
+                          other->overridden_target_name_)
+               : strcmp(overridden_target_name_,
+                        other->overridden_target_name_);
+  }
+
+  bool check_call_host(const char* host, grpc_auth_context* auth_context,
```

Comment:
This function and above cmp function has large code duplication. Either refactor into common function or add a TODO for future refactoring.

Reply:
cmp function has been refactored.